homeassistant/components/tesla/__init__.py

```diff
@@ -253,16 +303,21 @@ def device_info(self):
 
     async def async_added_to_hass(self):
         """Register state update callback."""
+        self.async_on_remove(self.coordinator.async_add_listener(self.refresh))
 
     async def async_will_remove_from_hass(self):
         """Prepare for unload."""
 
     async def async_update(self):
         """Update the state of the device."""
-        if self.controller.is_token_refreshed():
-            (refresh_token, access_token) = self.controller.get_tokens()
-            _async_save_tokens(
-                self.hass, self.config_entry, access_token, refresh_token
-            )
-            _LOGGER.debug("Saving new tokens in config_entry")
-        await self.tesla_device.async_update()
+        _LOGGER.debug("Updating state for: %s", self.name)
+        await self.coordinator.async_request_refresh()
+        self.refresh()
+
+    def refresh(self) -> None:
+        """Refresh the state of the device.
+
+        This assumes the coordinator has updated the controller.
+        """
+        self.tesla_device.refresh()
+        self.schedule_update_ha_state()
```

Comment:
We need to use the async api in async context. Use `self.async_write_ha_state`.

Reply:
This is a sync function intended to be used as a callback. Is this considered async context?

Reply:
Also this breaks the component on start up switching to `self.async_write_ha_state()`.
```text
2020-08-08 13:18:10 ERROR (MainThread) [homeassistant.components.binary_sensor] tesla: Error on device update!
 Traceback (most recent call last):
   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 320, in _async_add_entity
     await entity.async_device_update(warning=False)
   File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 463, in async_device_update
     await self.async_update()  # type: ignore
   File "/config/custom_components/tesla/__init__.py", line 317, in async_update
     self.refresh()
   File "/config/custom_components/tesla/__init__.py", line 325, in refresh
     self.async_write_ha_state()
   File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 286, in async_write_ha_state
     raise NoEntitySpecifiedError(
 homeassistant.exceptions.NoEntitySpecifiedError: No entity id specified for entity Tesla Model 3 charger sensor
```

Reply:
We should make it an async callback by decorating it with `@callback` imported from core.py.

I'm assuming `self.tesla_device.refresh` is async safe.

Reply:
Yes, but even with the callback, the NoEntitySpecifiedError occurs.

Reply:
Remove line 312.

Reply:
Done.